src/Zynga/Framework/PgData/V1/PgModel/Reader.hh

```diff
@@ -83,34 +86,44 @@ class Reader implements ReaderInterface {
         $row = $resultSet->get(0);
         if ($row instanceof PgRowInterface) {
           // fetchResultSetFromDatabase already writes the row to cache.
-          if($getLocked === false) {
-            $pgModel->cache()->unlockRowCache($row);
-          }
-          
+
           return $row;
         }
       }
-      
-      // 5) If we got more than 1 result,
-      // then release lock for all of them acquired by fetchResultSetFromDatabase
-      if($resultSet->count() > 1) {
-        foreach ($resultSet->toArray() as $resultObj) {
-          if (!$resultObj instanceof PgRowInterface) {
-            continue;
-          }
-          
-          $pgModel->cache()->unlockRowCache($resultObj);
-        }
-        
-        return null;
-      }
 
-      // 6) row not found by pk. Always unlock in this case
-      $pgModel->cache()->unlockRowCache($obj);
       return null;
 
     } catch (Exception $e) {
+      // unlock the result obj even asked for not to unlock when an exception happens
+      $getLocked = false;
       throw $e;
+    } finally {
+      // release all the possible locks
+      if ($pgModel !== null && $resultSet !== null) {
+        $cache = $pgModel->cache();
+        foreach ($resultSet->toArray() as $idx => $resultObj) {
+
+          if (!$resultObj instanceof PgRowInterface) {
+            continue;
+          }
+          if ($cache->getDataCache()->isLocked($resultObj) === true) {
+            // 6) Don't unlock if asked not to.
+            if ($idx === 0 && $getLocked === false) {
```

Comment:
You should always unlock when count > 1. In this case if $resultSet has two times, you wouldn't unlock index 0 but will unlock index 1. 

I would suggest separate this out of the for each and check for count of result set.